File: MinimumWindowSubstring.java
import java.util.*;

public class MinimumWindowSubstring {
    // doesn't pass one test case for some reason, can't debug
    public static String minWindow(String s, String t) {
        HashMap<Character, Integer> needMap = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            if (needMap.containsKey(t.charAt(i))) {
                needMap.put(t.charAt(i), needMap.get(t.charAt(i)) + 1);
            } else {
                needMap.put(t.charAt(i), 1);
            }
        }

        int needMapSize = needMap.size();

        HashMap<Character, Integer> haveMap = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            haveMap.put(t.charAt(i), 0);
        }
        String minSubString = null;
        int start = 0;
        int end = 0;
        int haveCharCount = 0;

        while (end < s.length()) {
            char currChar = s.charAt(end);
            if (haveMap.containsKey(currChar)) {
                haveMap.put(currChar, haveMap.get(currChar) + 1);
                if (haveMap.get(currChar) == needMap.get(currChar)) {
                    haveCharCount++;
                }
            }

            while (haveCharCount == needMapSize) {
                String subS = s.substring(start, end + 1);
                if (minSubString == null || subS.length() < minSubString.length()) {
                    minSubString = subS;
                }
                char leftChar = s.charAt(start);
                if (haveMap.containsKey(leftChar)) {
                    haveMap.put(leftChar, haveMap.get(leftChar) - 1);
                    if (haveMap.get(leftChar) < needMap.get(leftChar)) {
                        haveCharCount--;
                    }
                }
                start++;
            }
            end++;
        }

        if (minSubString == null) {
            return "";
        }

        return minSubString;
    }

    public static void main(String[] args) {

    }
}

File: MinimumWindowSubstring.py
def minWindow(s, t):
    """
    :type s: str
    :type t: str
    :rtype: str
    """

    needMap = {}
    for c in t:
        if c in needMap:
            needMap[c] += 1
        else:
            needMap[c] = 1
    
    needMapSize = len(needMap)

    haveMap = {}
    for c in t:
        haveMap[c] = 0
    
    minSubstring = None
    start = 0
    end = 0
    haveCharCount = 0

    while end < len(s):
        currChar = s[end]
        if currChar in haveMap:
            haveMap[currChar] += 1
            if needMap[currChar] == haveMap[currChar]:
                haveCharCount += 1
        while haveCharCount == needMapSize:
            subS = s[start : end + 1]
            if not minSubstring or len(subS) < len(minSubstring):
                minSubstring = subS
            leftChar = s[start]
            if leftChar in haveMap:
                haveMap[leftChar] -= 1
                if haveMap[leftChar] < needMap[leftChar]:
                    haveCharCount -= 1
            start += 1
        end += 1

    if not minSubstring:
        return ""
    return minSubstring

def main():
    s = "ADOBECODEBANC"
    t = "ABC"
    minWindow(s, t)

if __name__ == "__main__":
    main()